control-base/src/main/java/fi/nls/oskari/control/feature/SaveFeatureHandler.java

```diff
@@ -4,80 +4,73 @@
-                            "xmlns:ogc='http://www.opengis.net/ogc' " +
-                            "xmlns:wfs='http://www.opengis.net/wfs'>" +
-                            "<wfs:Update typeName='" + layer.getName() + "'>");
-            JSONArray jsonArray = jsonObject.getJSONArray("featureFields");
-            for (int i = 0; i < jsonArray.length(); i++) {
-                requestData.append("<wfs:Property><wfs:Name>" + jsonArray.getJSONObject(i).getString("key") +
-                        "</wfs:Name><wfs:Value>" + jsonArray.getJSONObject(i).getString("value") +
-                        "</wfs:Value></wfs:Property>");
-            }
+                if (!canEdit(layer, params.getUser())) {
+                    throw new ActionDeniedException("User doesn't have edit permission for layer: " + layer.getId());
+                }
+                ByteArrayOutputStream baos = new ByteArrayOutputStream();
 
-            if (jsonObject.has("geometries")) {
-                insertGeometries(lc.getGMLGeometryProperty(), requestData, jsonObject.getJSONObject("geometries"), srsName);
-            }
-            requestData.append("<ogc:Filter><ogc:FeatureId fid='" + jsonObject.getString("featureId") + "'/></ogc:Filter></wfs:Update></wfs:Transaction>");
+                Feature feature = getFeature(featureJSON);
+                FeatureWFSTRequestBuilder.updateFeature(baos, feature);
 
-            String responseString = postPayload(layer, requestData.toString());
-            flushLayerTilesCache(layer.getId());
+                final String wfstMessage = baos.toString();
+                LOG.debug("Updating feature to service at", layer.getUrl(), "with payload", wfstMessage);
 
-            if (responseString.indexOf("Exception") > -1) {
-                ResponseHelper.writeResponse(params, "Exception");
-            } else if (responseString.indexOf("<wfs:totalUpdated>1</wfs:totalUpdated>") > -1) {
-                ResponseHelper.writeResponse(params, "");
+                String responseString = postPayload(layer, wfstMessage);
+                flushLayerTilesCache(layer.getId());
+
+                if (responseString.indexOf("Exception") > -1) {
+                    exceptions.add(responseString);
+                }
             }
-        } catch (JSONException ex) {
-            log.error(ex, "JSON processing error");
-            throw new ActionException("JSON processing error", ex);
-        }
-    }
 
-    protected void insertGeometries(String geometryProperty, StringBuilder requestData, JSONObject geometries, String srsName) throws ActionParamsException, JSONException {
-        String geometryType = geometries.getString("type");
-        if(!isAllowedGeomType(geometryType)) {
-            throw new ActionParamsException("Invalid geometry type: " + geometryProperty);
+            if(exceptions.size() > 0 ) {
+                throw new ActionException("Cannot save feature(s): " + exceptions.toString());
```

Comment:
This is confusing since we might have successfully updated multiple features but one more more failed.